src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentTrainEnumerator.cs

```diff
@@ -0,0 +1,96 @@
+    private readonly TransactionOperationContext _indexContext;
+    private readonly Index _index;
+    private readonly IndexType _indexType;
+    private readonly HashSet<string> _collections;
+    private readonly int _take;
+    private IEnumerator<LazyStringValue> _itemsEnumerable;
+
+    public CoraxDocumentTrainEnumerator(TransactionOperationContext indexContext, Index index, IndexType indexType, DocumentsStorage storage, QueryOperationContext queryContext, HashSet<string> collections, int take = int.MaxValue)
+    {
+        _indexContext = indexContext;
+        _index = index;
+        _indexType = indexType;
+        _take = take;
+
+        _documentStorage = storage;
+        _queryContext = queryContext;
+        _collections = collections;
+    }
+
+    private IEnumerable<LazyStringValue> GetItems()
+    {
+        var scope = new IndexingStatsScope(new IndexingRunStats());
+        foreach (var collection in _collections)
+        {
+            using var itemEnumerator = _index.GetMapEnumerator(GetItemsEnumerator(_queryContext, collection, _take), collection, _indexContext, scope, _indexType);
+            while (true)
+            {
+                if (itemEnumerator.MoveNext(_queryContext.Documents, out var _, out var _) == false)
+                    break;
+
+                Document doc = itemEnumerator.Current.Item as Document;
+                if (doc == null)
+                    continue;
+
+                var reader = doc.Data;
+                int properties = reader.Count;
+                BlittableJsonReaderObject.PropertyDetails details = default;
+                for (int i = 0; i < properties; i++)
+                {
+                    reader.GetPropertyByIndex(i, ref details);
+                    switch (details.Token)
+                    {
+                        case BlittableJsonToken.String:
+                            yield return (LazyStringValue)details.Value;
+                            break;
+                        case BlittableJsonToken.CompressedString:
+                            yield return ((LazyCompressedStringValue)details.Value).ToLazyStringValue();
+                            break;
+                    }
+                }
+            }
+        }
+    }
+
+    private IEnumerable<Document> GetDocumentsEnumerator(QueryOperationContext queryContext, string collection, long take = int.MaxValue)
+    {
+        if (collection == Constants.Documents.Collections.AllDocumentsCollection)
+            return _documentStorage.GetUniformlyDistributedDocumentsFrom(queryContext.Documents, take);
+        return _documentStorage.GetUniformlyDistributedDocumentsFrom(queryContext.Documents, collection, take);
+    }
+
+    private IEnumerable<IndexItem> GetItemsEnumerator(QueryOperationContext queryContext, string collection, long take = int.MaxValue)
+    {
+        foreach (var document in GetDocumentsEnumerator(queryContext, collection, take))
```

Comment:
What if an index is defined on counters or time series (not on documents)?

CC @ayende

Reply:
That is something to take into account.
I'm fine with not having (at least for 6.0) the best dictionary for that)

Reply:
Agree.

Reply:
So maybe we should skip it completely for indexes on counters and TS?

Reply:
Are collections and their data involved on such indexes?

Reply:
We operate on counters, time-series data per collection then. See here:

- counters
https://github.com/ravendb/ravendb/blob/05e95d4bf2ca6c9db8515190f57ca113511b48f0/src/Raven.Server/Documents/Indexes/Workers/Counters/MapCounters.cs#L18-L31

- time series

https://github.com/ravendb/ravendb/blob/75949c69d610d26dc5407f80a5bb84e5825ec5e3/src/Raven.Server/Documents/Indexes/Workers/TimeSeries/MapTimeSeries.cs#L19-L33

Reply:
What are we gonna do with that? Please note that the index enumerator expects to get items of a specified type. So effectively this won't return anything if an index is defined on counters while you provide documents:

```
using var itemEnumerator = _index.GetMapEnumerator(GetItemsEnumerator(_queryContext, collection, _take), collection, _indexContext, scope, _indexType);
```

(it maybe it will error / crash - not sure)